AutoHeuristic: Enable explicit support for ranking

AlnisM opened the pull request:
This PR adds support for heuristics that rank choices in AutoHeuristic. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133611
* #133608
* #131714
* __->__ #131710
* #131705



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

AlnisM committed:
AutoHeuristic: Enable explicit support for ranking

[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -505,7 +515,7 @@ def try_heuristic(m, n, k, choices, mat1, mat2, mat2_dtype, layout):
 
 def mm_autoheuristic(
-    mat1, mat2, m, n, k, choices, name, input_nodes, ops, precondition
+    mat1, mat2, m, n, k, choices, name, input_nodes, ops, precondition, top_k=-1
```
nit: would it make more sense to make `top_k` an `Optional[int]` ?

eellison commented on torch/_inductor/autoheuristic/autoheuristic.py:
```diff
@@ -136,6 +136,20 @@ def get_choice(self) -> Choice:
                 return decision
         return self.fallback()
 
+    def get_top_k_choices(self, top_k: int) -> Optional[List[Choice]]:
```
I guess eventually, we could adjust the top_k # to benchmark based on likelihood we think of additional choice being chosen

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Enable explicit support for ranking"

This PR adds support for heuristics that rank choices in AutoHeuristic. 




[ghstack-poisoned]

AlnisM commented:
@pytorchbot merge
